pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java

```diff
@@ -71,12 +73,41 @@ protected void scheduleReadOnActiveConsumer() {
             havePendingRead = false;
         }
 
-        // When a new consumer is chosen, start delivery from unacked message. If there is any pending read operation,
-        // let it finish and then rewind
-        if (!havePendingRead) {
+        if (havePendingRead) {
+            return;
+        }
+
+        // When a new consumer is chosen, start delivery from unacked message.
+        // If there is any pending read operation, let it finish and then rewind
+
+        if (subscriptionType != SubType.Failover || serviceConfig.getReadOnActiveConsumerDelayTimeMillis() <= 0) {
+            if (log.isDebugEnabled()) {
+                log.debug("[{}] Rewind cursor and read more entries without delay", name);
+            }
             cursor.rewind();
             readMoreEntries(ACTIVE_CONSUMER_UPDATER.get(this));
+            return;
         }
+
+        // If subscription type is Failover, delay rewinding cursor and
+        // reading more entries in order to prevent message duplication
+
+        if (readOnActiveConsumerTask != null && !readOnActiveConsumerTask.isDone()) {
+            readOnActiveConsumerTask.cancel(false);
+            if (!readOnActiveConsumerTask.isCancelled()) {
+                return;
+            }
+        }
+
+        if (log.isDebugEnabled()) {
```

Comment:
can we move this debug log in `scheduled-task` at line 108

Reply:
Fixed.